pkg/controller/common/ignition_converter.go

```diff
@@ -0,0 +1,332 @@
+	// Convert Converts the given Ignition config instance to the target one (designated by the target version
+	// of the Versions tuple). If the conversion fails or the source type is not the expected one an error
+	// is returned.
+	Convert(source any) (any, error)
+
+	// Versions returns the versions conversion tuple the converter implements.
+	Versions() conversionTuple
+}
+
+type baseConverter struct {
+	tuple conversionTuple
+}
+
+// converterFn represents a conversion function that may fail
+type converterFn func(source any) (any, error)
+
+// converterFn represents a generic typed conversion function that may fail
+type converterTypedFn[S any, T any] func(source S) (T, error)
+
+// converterFn represents a generic typed conversion function that always succeeds
+type converterTypedNoErrorFn[S any, T any] func(source S) T
+
+type functionConverter struct {
+	baseConverter
+	conversionFunc converterFn
+}
+
+// newFunctionConverter builds a function based converter for the given sourceVersion and targetVersion
+func newFunctionConverter(sourceVersion, targetVersion semver.Version, conversionFunc converterFn) *functionConverter {
+	return &functionConverter{
+		baseConverter: baseConverter{
+			tuple: conversionTuple{sourceVersion: sourceVersion, targetVersion: targetVersion},
+		},
+		conversionFunc: conversionFunc,
+	}
+}
+
+// newFunctionConverterTyped builds an Ignition version converter for the given sourceVersion and targetVersion,
+// based on a generic function.
+func newFunctionConverterTyped[S any, T any](sourceVersion, targetVersion semver.Version, conversionFunc converterTypedFn[S, T]) *functionConverter {
+	return newFunctionConverter(
+		sourceVersion, targetVersion,
+		func(source any) (any, error) {
+			if input, ok := source.(S); ok {
+				return conversionFunc(input)
+			}
+			var empty T
+			return empty, errIgnitionConverterWrongSourceType
+		})
+}
+
+// newFunctionConverterTypedNoError builds an Ignition version converter for the given sourceVersion and targetVersion,
+// based on a generic error-free conversion function.
+func newFunctionConverterTypedNoError[S any, T any](sourceVersion, targetVersion semver.Version, conversionFunc converterTypedNoErrorFn[S, T]) *functionConverter {
+	return newFunctionConverterTyped(sourceVersion, targetVersion, func(s S) (T, error) { return conversionFunc(s), nil })
+}
+
+func (c *functionConverter) Convert(source any) (any, error) {
+	result, err := c.conversionFunc(source)
+	if err != nil && errors.Is(err, errIgnitionConverterWrongSourceType) {
+		return result, fmt.Errorf("conversion from %v to %v failed: %w", c.tuple.sourceVersion, c.tuple.targetVersion, err)
+	}
+	return result, err
+}
+
+func (c *functionConverter) Versions() conversionTuple {
+	return c.tuple
+}
+
+// buildConverterList creates a list that contains all the MCO supported [ignitionVersionConverter] for each
+// of the supported version maps.
+func buildConverterList() []ignitionVersionConverter {
```

Comment:
Curious where you built this list from, for example, it seems that a 3.1 config can't be translated to anything else?

Also if I'm understanding this correctly, this essentially defines all supported ignition versions and ignition type conversions? What happens if a type translation (e.g. 3.5->2.2) is requested somehow?

Reply:
Thanks for the detailed answers!

> I think there's no way right now to receive an up-conversion request from 3.1 to i.e. 3.5.

Ah, I think I may be mis-remembering how the conversion -> generate rendered config works. Looks like if we get a v3.x config we just cast it upwards to 3.5 directly, since we don't remove any fields from the ignition spec, that generally works. I had initially thought that we had to actually do a translation which isn't the case, so that is indeed not needed.

(As a note, it looks like the openshift-docs does still contain references to e.g. 3.1.0 ignition https://github.com/search?q=repo%3Aopenshift%2Fopenshift-docs%203.1.0&type=code and while that's harmless we should probably still try to keep things up to date not to confuse people)